src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -228,6 +267,186 @@ TimeSynchronizationServer & TimeSynchronizationServer::Instance()
     return sTimeSyncInstance;
 }
 
+TimeSynchronizationServer::TimeSynchronizationServer() :
+#if TIME_SYNC_ENABLE_TSC_FEATURE
+    mOnDeviceConnectedCallback(OnDeviceConnectedWrapper, this),
+    mOnDeviceConnectionFailureCallback(OnDeviceConnectionFailureWrapper, this),
+#endif
+    mOnTimeSyncCompletion(OnTimeSyncCompletionWrapper, this), mOnFallbackNTPCompletion(OnFallbackNTPCompletionWrapper, this)
+{}
+
+void TimeSynchronizationServer::AttemptToGetFallbackNTPTimeFromDelegate()
+{
+    // Sent as a char-string to the delegate so they can read it easily
+    char defaultNTP[kMaxDefaultNTPSize];
+    MutableCharSpan span(defaultNTP);
+    if (GetDefaultNtp(span) != CHIP_NO_ERROR)
+    {
+        emitTimeFailureEvent(kRootEndpointId);
+        return;
+    }
+    if (span.size() > kMaxDefaultNTPSize)
+    {
+        emitTimeFailureEvent(kRootEndpointId);
+        return;
+    }
+    if (GetDelegate()->UpdateTimeUsingNTPFallback(span, &mOnFallbackNTPCompletion) != CHIP_NO_ERROR)
+    {
+        emitTimeFailureEvent(kRootEndpointId);
+    }
+}
+
+#if TIME_SYNC_ENABLE_TSC_FEATURE
+void TimeSynchronizationServer::OnDeviceConnectedFn(Messaging::ExchangeManager & exchangeMgr, const SessionHandle & sessionHandle)
+{
+    // Connected to our trusted time source, let's read the time.
+    app::AttributePathParams readPaths[2];
+    readPaths[0] = app::AttributePathParams(kRootEndpointId, app::Clusters::TimeSynchronization::Id,
+                                            app::Clusters::TimeSynchronization::Attributes::UTCTime::Id);
+    readPaths[1] = app::AttributePathParams(kRootEndpointId, app::Clusters::TimeSynchronization::Id,
+                                            app::Clusters::TimeSynchronization::Attributes::Granularity::Id);
+
+    app::InteractionModelEngine * engine = app::InteractionModelEngine::GetInstance();
+    app::ReadPrepareParams readParams(sessionHandle);
+    readParams.mpAttributePathParamsList    = readPaths;
+    readParams.mAttributePathParamsListSize = 2;
+
+    auto attributeCache = Platform::MakeUnique<app::ClusterStateCache>(*this);
```

Comment:
Why do you need a ClusterStateCache?  That's a bunch of code being pulled in with a bunch of codesize overhead that's totally not needed if we're just trying to read these two attributes....

The ClusterStateCache is written with high-resource controllers in mind, not constrained devices.

Reply:
Addressed in https://github.com/project-chip/connectedhomeip/pull/29396